### PR TITLE
Add rules `order-in-routes` and `order-in-controllers`

### PR DESCRIPTION
**Pull request by michalsnik**
This PR adds following rules:
- order-in-routes
- order-in-controllers

according to proposal: #7 

### order-in-routes
You should write code grouped and ordered in this way:
1. Services
2. Default route's properties
3. Custom properties
4. model() hook
5. Other route's methods (beforeModel etc.)
6. Actions
7. Custom / private methods

```javascript
const { Route, inject: { service }, get } = Ember;

export default Route.extend({
  // 1. Services
  currentUser: service(),

  // 2. Default route's properties
  queryParams: {
    sortBy: { refreshModel: true },
  },

  // 3. Custom properties
  customProp: 'test',

  // 4. Model hook
  model() {
    return this.store.findAll('article');
  },

  // 5. Other route's methods
  beforeModel() {
    if (!get(this, 'currentUser.isAdmin')) {
      this.transitionTo('index');
    }
  },

  // 6. All actions
  actions: {
    sneakyAction() {
      return this._secretMethod();
    },
  },

  // 7. Custom / private methods
  _secretMethod() {
    // custom secret method logic
  },
});
```

### order-in-controllers
You should write code grouped and ordered in this way:
1. Services
2. Default controller's properties
3. Custom properties
4. Single line computed properties
5. Multi line computed properties
6. Observers
7. Actions
8. Custom / private methods

```javascript
const { Controller, computed, inject: { service }, get } = Ember;

export default Controller.extend({
  // 1. Services
  currentUser: service(),

  // 2. Default route's properties
  queryParams: ['view'],

  // 3. Custom properties
  attitude: 10,

  // 4. Single line Computed Property
  health: alias('model.health'),

  // 5. Multiline Computed Property
  levelOfHappiness: computed('attitude', 'health', function() {
    return get(this, 'attitude') * get(this, 'health') * Math.random();
  }),

  // 6. Observers
  onVahicleChange: observer('vehicle', function() {
    // observer logic
  }),

  // 7. All actions
  actions: {
    sneakyAction() {
      return this._secretMethod();
    },
  },

  // 8. Custom / private methods
  _secretMethod() {
    // custom secret method logic
  },
});
```

### CODE REVIEW

File: rules/order-in-components.js [hunk 1]
```diff
@@ -110,12 +27,77 @@ module.exports = function(context) {
...
+function isCustomFunction(property) {
+  return (
+    utils.isFunctionExpression(property.value) ||
+    utils.isCallWithFunctionExpression(property.value)
+  ) && !ember.isComponentLifecycleHookName(property.key.name);
+};
 
+function isLifecycleHook(property) {
+  return utils.isFunctionExpression(property.value) &&
+    ember.isComponentLifecycleHookName(property.key.name);
```
Comment (bardzusny) [inline rules/order-in-components.js]: I'm quite certain that you shouldn't repeat definitions of all these functions in each rule, rather isolate it to shared util.

File: rules/order-in-components.js [hunk 2]
```diff
@@ -110,12 +27,77 @@ module.exports = function(context) {
...
+function isActionsProp(property) {
+  return property.key.name === 'actions' && utils.isObjectExpression(property.value);
+};
+
+function isCustomFunction(property) {
+  return (
+    utils.isFunctionExpression(property.value) ||
+    utils.isCallWithFunctionExpression(property.value)
+  ) && !ember.isComponentLifecycleHookName(property.key.name);
+};
```
Comment (bardzusny) [inline rules/order-in-components.js]: `;` - not really here.

File: rules/order-in-components.js [hunk 3]
```diff
@@ -110,12 +27,77 @@ module.exports = function(context) {
...
 
       if (unorderedProperty) {
         report(unorderedProperty.node);
       }
     }
   };
+};
+
+function getOrderValue(property) {
+  var val = null;
```
Comment (bardzusny) [inline rules/order-in-components.js]: Stylistic issue only, but I can't see any reason for this variable - why not simply `return` statement in each `if`/`else if` block?

File: rules/order-in-controllers.js [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+
+var utils = require('./utils/utils');
+var ember = require('./utils/ember');
+
+//------------------------------------------------------------------------------
+// Organizing - Organize your routes and keep order in objects
+//------------------------------------------------------------------------------
+
+module.exports = function(context) {
+
```
Comment (bardzusny) [inline rules/order-in-controllers.js]: Padded block - I'd set-up basic `.eslintrc` even for non-transpiled old-fashioned ES5 code, by the way (maybe no in this PR, though).

File: rules/order-in-controllers.js [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+function isActionsProp(property) {
+  return property.key.name === 'actions' && utils.isObjectExpression(property.value);
+}
+
+function isCustomFunction(property) {
+  return (
+    utils.isFunctionExpression(property.value) ||
+    utils.isCallWithFunctionExpression(property.value)
+  );
+}
```
Comment (bardzusny) [inline rules/order-in-controllers.js]: Basically copy-paste of helpers in `rules/order-in-components.js` - they should be quite easy to isolate & share code.

File: rules/order-in-components.js (hunk 3 above)
Comment (michalsnik) [inline rules/order-in-components.js]: I like to keep as few return statements in functions as possible. In this scenario this doesn't improve anything, but for larger functions I think it's a good practice.

File: rules/order-in-controllers.js (hunk 4 above)
Comment (michalsnik) [inline rules/order-in-controllers.js]: Yup, agree and this is something that I will consider in further PR's

File: rules/order-in-components.js (hunk 1 above)
Comment (michalsnik) [inline rules/order-in-components.js]: Ok, I moved repeated code to utils

File: rules/order-in-components.js [hunk 6]
```diff
@@ -110,12 +27,69 @@ module.exports = function(context) {
...
+function isActionsProp(property) {
+  return ember.isActionsProp(property);
+}
+
+function isCustomFunction(property) {
+  return ember.isFunctionExpression(property.value) &&
+    !ember.isComponentLifecycleHookName(property.key.name);
+}
+
+function isLifecycleHook(property) {
```
Comment (bardzusny) [inline rules/order-in-components.js]: I'd isolate these composed functions to `ember` utility, as well.

File: rules/order-in-components.js [hunk 7]
```diff
@@ -110,12 +27,69 @@ module.exports = function(context) {
...
+
+function isObserverProp(property) {
+  return ember.isObserverProp(property.value);
+}
+
+function isActionsProp(property) {
+  return ember.isActionsProp(property);
+}
+
+function isCustomFunction(property) {
```
Comment (bardzusny) [inline rules/order-in-components.js]: I'd isolate these composed functions to ember utility, as well.

File: rules/order-in-components.js [hunk 8]
```diff
@@ -110,12 +27,69 @@ module.exports = function(context) {
...
+  } else if (isActionsProp(property)) {
+    val = 70;
+  } else if (isCustomFunction(property)) {
+    val = 80;
+  }
+
+  return val;
+}
+
+function isAnyProp(property) {
```
Comment (bardzusny) [inline rules/order-in-components.js]: These wrapping functions seem unnecessary to me - `ember` util seems like perfect place for them, to be used directly.

File: rules/utils/utils.js [hunk 9]
```diff
@@ -196,3 +197,19 @@ function parseArgs(node) {
...
+
+/**
+ * Find property that is in wrong order
+ *
+ * @param  {Array} arr Properties with their order value
+ * @return {Object}    Unordered property or null
+ */
+function findUnorderedProperty(arr) {
+  for(var i = 1; i < arr.length; i++) {
+    if(arr[i-1].order > arr[i].order) {
```
Comment (bardzusny) [inline rules/utils/utils.js]: Again lint-level thing: space after `if` (used everywhere else). Similarly applies to `for` expression line above, I can presume.

File: rules/utils/utils.js [hunk 10]
```diff
@@ -196,3 +197,19 @@ function parseArgs(node) {
...
 }
+
+/**
+ * Find property that is in wrong order
+ *
+ * @param  {Array} arr Properties with their order value
+ * @return {Object}    Unordered property or null
+ */
+function findUnorderedProperty(arr) {
+  for(var i = 1; i < arr.length; i++) {
```
Comment (bardzusny) [inline rules/utils/utils.js]: Can't we start with 0 and just do `+1` where necessary? Just stylistic issue, feels weird to me to pretend that index starts at `1` (even though it's not the intention here).

File: test/order-in-controllers.js [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+      }],
+    },
+    {
+      code: 'export default Controller.extend({test: "asd", obs: observer("asd", function() {}), comp: computed("asd", function() {}), actions: {} });',
+      parserOptions: {ecmaVersion: 6, sourceType: "module"},
+      errors: [{
+        message: 'Check order of properties',
+      }],
+    },
+  ]
```
Comment (bardzusny) [inline test/order-in-controllers.js]: Another lint-level issue - inconsistency with trailing comma (across whole PR, actually).

File: rules/utils/utils.js (hunk 10 above)
Comment (michalsnik) [inline rules/utils/utils.js]: It's either `i = 1; i < arr.length; i++` or `i = 0; i < arr.length - 1; i++`. I can change it, but it's rather personal preference.

File: test/order-in-controllers.js (hunk 11 above)
Comment (michalsnik) [inline test/order-in-controllers.js]: I'll leave it as is. Later on I will add Eslint to assure consistency ;)

File: rules/order-in-components.js (hunk 8 above)
Comment (michalsnik) [inline rules/order-in-components.js]: Hmm, maybe you're right. I moved all logic to utilities now.

**Pull request merged**